stl/inc/xutility

```diff
@@ -4043,69 +4043,105 @@ inline constexpr unreachable_sentinel_t unreachable_sentinel{};
+template <class _Iter1, class _Iter2>
+_INLINE_VAR constexpr bool _Iterators_are_contiguous = contiguous_iterator<_Iter1> //
+    && contiguous_iterator<_Iter2>;
+#else // ^^^ defined(__cpp_lib_concepts) ^^^ / vvv !defined(__cpp_lib_concepts) vvv
+// When concepts aren't available, we can detect pointers. (Iterators should be unwrapped before using this.)
+template <class _Iter1, class _Iter2>
+_INLINE_VAR constexpr bool _Iterators_are_contiguous = conjunction_v<is_pointer<_Iter1>, is_pointer<_Iter2>>;
+#endif // ^^^ !defined(__cpp_lib_concepts) ^^^
 
-template <class _Elem>
-struct _Ptr_cat_helper<_Elem, _Elem> { // determines _Ptr_cat's result when the types are the same
-    static constexpr bool _Really_trivial     = is_trivial_v<_Elem>;
-    static constexpr bool _Trivially_copyable = is_trivially_copyable_v<_Elem>;
+// _Can_memcopy_elements<_Ty1, _Ty2> reports whether `_Ty1` can be memcopy'd into  _Ty2`.
+// Here, _Ty1 and _Ty2 aren't top-level const, because we remove_const_t before using _Can_memcopy_elements.
+// There are different specialization wrt move / copy and whether the destination is uninitialized memory
+enum class _Memcopy_cat {
+    _Copy,
+    _Copy_uninitialized,
+    _Move,
+    _Move_uninitialized,
 };
 
-template <class _Anything>
-struct _Ptr_cat_helper<_Anything*, const _Anything*> {
-    // determines _Ptr_cat's result when all we do is add const to a pointer
-    static constexpr bool _Really_trivial     = true;
-    static constexpr bool _Trivially_copyable = true;
-};
+// Integral types are eligible for memcopy in very specific cases.
+// * Enumerations are treated as their underlying integral type
+// * They must be the same size. (`int == long` is eligible; `int == long long` isn't.)
+// * They are either both bool or both non bool
+template <class _Source, class _Dest>
+_INLINE_VAR constexpr bool _Memcopy_matching_integral = sizeof(_Source) == sizeof(_Dest) //
+                                                     && is_same_v<bool, _Source> == is_same_v<bool, _Dest> //
+                                                     && is_integral_v<_Source>&& is_integral_v<_Dest>;
 
-template <class _Anything>
-struct _Ptr_cat_helper<_Anything*, volatile _Anything*> {
-    // determines _Ptr_cat's result when all we do is add volatile to a pointer
-    static constexpr bool _Really_trivial     = true;
-    static constexpr bool _Trivially_copyable = true;
-};
 
-template <class _Anything>
-struct _Ptr_cat_helper<_Anything*, const volatile _Anything*> {
-    // determines _Ptr_cat's result when all we do is add cv to a pointer
-    static constexpr bool _Really_trivial     = true;
-    static constexpr bool _Trivially_copyable = true;
-};
+// Matching integer type are eligible for memcopy
+template <class _Source, class _Dest, _Memcopy_cat _Cat,
+    bool = _Memcopy_matching_integral<_Unwrap_enum_t<_Source>, _Unwrap_enum_t<_Dest>>>
+_INLINE_VAR constexpr bool _Can_memcopy_elements = true;

```

Comment:
I think one could improve this by reusing the result from `_Memcopy_matching_integral`, aka
```suggestion
template <class _Source, class _Dest, _Memcopy_cat _Cat,
    bool _Result = _Memcopy_matching_integral<_Unwrap_enum_t<_Source>, _Unwrap_enum_t<_Dest>>>
_INLINE_VAR constexpr bool _Can_memcopy_elements = _Result;
```

This would make the `// Arbitrary types are not eligible` specialization obsolete.

That said it would be harder to find the default case